File: songs/python/22-unique-b.py
from MySongs import songs

tags = [
  tag for song in songs
  if 'tags' in song
    for tag in song['tags']
]

def exclude(val, tags):
  return [tag for tag in tags if tag != val]

def unique(tags):
  if len(tags) <= 1: return tags
  else:
    head, *tail = tags
    return [head] + unique(exclude(head, tail))

print(unique(tags))

File: songs/ruby/22-unique-b.rb
require_relative 'MySongs'
include Songs

def exclude(val, tags)
  tags.map{ |tag| tag unless tag == val }
    .compact
end

def unique(tags)
  if tags.length() <= 1 then
    tags
  else
    head, *tail = tags
    return [head] + unique(exclude(head, tail))
  end
end

tags = SONGS
  .map { |song| song['tags'] }
  .compact.flatten

puts "#{unique(tags)}"
